erts/emulator/beam/erl_db.c

```diff
@@ -2136,8 +2150,14 @@ BIF_RETTYPE ets_internal_delete_all_2(BIF_ALIST_2)
     DB_BIF_GET_TABLE(tb, DB_WRITE, LCK_WRITE, BIF_ets_internal_delete_all_2);
 
     if (BIF_ARG_2 == am_undefined) {
-        nitems = erts_make_integer(erts_atomic_read_nob(&tb->common.nitems),
-                                   BIF_P);
+        if(erts_flxctr_suspend_until_thr_prg_if_snapshot_ongoing(&tb->common.counters, BIF_P)){
+            db_unlock(tb, LCK_WRITE);
+            BIF_TRAP2(bif_export[BIF_ets_internal_delete_all_2], BIF_P, BIF_ARG_1, BIF_ARG_2);
```

Comment:
Yes, it is a pity but I don't have any good idea for how to get around this waiting unfortunately. We could remove the reading of the `nitems` counter here in case of ets:delete_all_objects but waiting or some other solution is still necessary because the counter is reset [here][1].

[1]: https://github.com/erlang/otp/blob/a45ea2655b661975668b1c26de4e23235ecf60e3/erts/emulator/beam/erl_db_catree.c#L2092

Reply:
This waiting is completely avoided with commit e7eef34422d33fb0ff62fde0984b8a30d2cc1572 that has been added to this pull request. In e7eef34422d33fb0ff62fde0984b8a30d2cc1572, the CA tree's db_delete_all_objects function decrements the nitems counter while deleting objects and stores the final number of deleted objects in a place that can be accessed by the caller of ets_internal_delete_all_2.